quantify the relationship between stocks by calculating the covariance between securities

larger house , more expensive--use correlation to measure the relationship between House price and house size

 For x, y vairables, covar>0,same direction;<0,opposite direction;=0, independent

location, year of construction, size-->determine house price

In [13]:
import numpy as np
import pandas as pd
from pandas_datareader import data as wb
import matplotlib.pyplot as plt

In [17]:
tickers=['PG','BEI.DE']
sec_data=pd.DataFrame()
for t in tickers:
    sec_data[t]=wb.DataReader(t,data_source='yahoo',start='2016-1-1')['Adj Close']
    

In [18]:
sec_data.tail()

PG     BEI.DE
Date                             
2020-06-01  117.250000        NaN
2020-06-02  118.059998  96.180000
2020-06-03  118.529999  98.959999
2020-06-04  116.050003  99.500000
2020-06-05  118.330002  99.080002

In [19]:
sec_returns=np.log(sec_data/sec_data.shift(1))

In [20]:
sec_returns

PG    BEI.DE
Date                          
2016-01-04       NaN       NaN
2016-01-05  0.003185  0.010790
2016-01-06 -0.009714 -0.006484
2016-01-07 -0.008772 -0.024730
2016-01-08 -0.015802 -0.002520
2016-01-11  0.009172 -0.008869
2016-01-12 -0.002089  0.029714
2016-01-13 -0.008664  0.006280
2016-01-14  0.003947 -0.023601
2016-01-15 -0.015484 -0.013922
2016-01-19  0.023071  0.029883
2016-01-20 -0.003122 -0.028482
2016-01-21  0.011668  0.010382
2016-01-22  0.008307  0.025249
2016-01-25 -0.006614  0.007708
2016-01-26  0.025184  0.006923
2016-01-27 -0.000127  0.011911
2016-01-28  0.012861 -0.016157
2016-01-29  0.023157  0.033585
2016-02-01 -0.007002 -0.000941
2016-02-02 -0.011157  0.006567
2016-02-03  0.010910 -0.007626
2016-02-04 -0.004944 -0.023836
2016-02-05  0.006177 -0.022319
2016-02-08  0.017337 -0.026495
2016-02-09  0.000242  0.002909
2016-02-10 -0.012419  0.000631
2016-02-11 -0.021299 -0.025175
2016-02-12  0.013550  0.014391
2016-02-16  0.005909  0.039029
...              ...       ...
2020-04-24 -0.005206 -0.006311
2020-04-27 -0.011260 -0.003594
2020-04-28 -0.004779  0.004859
2020-04-29  0.001624  0.002315
2020-04-30  0.006725  0.012211
2020-05-01 -0.008948       NaN
2020-05-04 -0.009029       NaN
2020-05-05  0.002071 -0.028924
2020-05-06 -0.025404  0.019082
2020-05-07 -0.008257  0.013534
2020-05-08  0.033143  0.021875
2020-05-11 -0.005535  0.004234
2020-05-12 -0.006613  0.005477
2020-05-13 -0.005515 -0.028339
2020-05-14 -0.000966 -0.013490
2020-05-15  0.007005  0.012842
2020-05-18  0.013864  0.033180
2020-05-19 -0.032979 -0.025211
2020-05-20  0.007443  0.005136
2020-05-21 -0.014762 -0.016789
2020-05-22  0.008741 -0.001738
2020-05-26 -0.005075 -0.005232
2020-05-27  0.016466  0.025252
2020-05-28  0.018874  0.017954
2020-05-29 -0.001207 -0.013488
2020-06-01  0.011408       NaN
2020-06-02  0.006885       NaN
2020-06-03  0.003973  0.028494
2020-06-04 -0.021145  0.005442
2020-06-05  0.019456 -0.004230

[1114 rows x 2 columns]

# Correlation and Covariance

In [24]:
PG_var=sec_returns['PG'].var()
PG_var

0.00016829759183846428

In [25]:
BEI_var=sec_returns['BEI.DE'].var()
BEI_var

0.00014733876440402706

In [26]:
PG_var_a=sec_returns['PG'].var()*250
PG_var_a

0.042074397959616074

In [28]:
BEI_var_a=sec_returns['BEI.DE'].var()*250
BEI_var_a

0.03683469110100677

In [35]:
cov_matrix=sec_returns.cov()   ##pandas.DataFrame.cov() --computes pairwise covariance of columns
cov_matrix

PG    BEI.DE
PG      0.000168  0.000050
BEI.DE  0.000050  0.000147

In [36]:
cov_matrix_a=sec_returns.cov()*250
cov_matrix_a

PG    BEI.DE
PG      0.042074  0.012398
BEI.DE  0.012398  0.036835

Explain:the variance of P&G stock is 0.042074,
the variance of BEI.DE is 0.036835
the co-variance of  BEI.DE and P&G is 0.012398

In [42]:
corr_matrix=sec_returns.corr() ###pandas.DataFrame.corr() --computes pairwise correlation of columns
corr_matrix

## we don't need to annualize the cprrelation table since it only shows the relationship between the 2 variable

PG    BEI.DE
PG      1.000000  0.316634
BEI.DE  0.316634  1.000000

##corr(prices) --focus on stock price levels
##corr(returns) --reflects the dependence between prices at different times and focuses on the returns of the portfolio

Explain: the stock returns of these 2 companies are weakly correlated